tensorboard/manager_e2e_test.py

```diff
@@ -0,0 +1,302 @@
+    # cycled exactly through the whole PID space. Neither Linux nor
+    # macOS recycles PIDs, so we should be fine.
+    self.assertNotEqual(r1.info.pid, r2.info.pid)
+    self._assert_live(r2.info, expected_logdir="./logs")
+
+  def test_exit_failure(self):
+    if os.name == "nt":
+      # TODO(@wchargin): This could in principle work on Windows.
+      self.skipTest("Requires a POSIX shell for the stub script.")
+    self._stub_tensorboard(
+        name="fail-with-77",
+        program=textwrap.dedent(
+            r"""
+            #!/bin/sh
+            printf >&2 'fatal: something bad happened\n'
+            printf 'also some stdout\n'
+            exit 77
+            """.lstrip(),
+        ),
+    )
+    start_result = manager.start(["--logdir=./logs", "--port=0"])
+    self.assertIsInstance(start_result, manager.StartFailed)
+    self.assertEqual(
+        start_result,
+        manager.StartFailed(
+            exit_code=77,
+            stderr="fatal: something bad happened\n",
+            stdout="also some stdout\n",
+        ),
+    )
+    self.assertEqual(manager.get_all(), [])
+
+  def test_exit_success(self):
+    # TensorBoard exiting with success but not writing the info file is
+    # still a failure to launch.
+    if os.name == "nt":
+      # TODO(@wchargin): This could in principle work on Windows.
+      self.skipTest("Requires a POSIX shell for the stub script.")
+    self._stub_tensorboard(
+        name="fail-with-0",
+        program=textwrap.dedent(
+            r"""
+            #!/bin/sh
+            printf >&2 'info: something good happened\n'
+            printf 'also some standard output\n'
+            exit 0
+            """.lstrip(),
+        ),
+    )
+    start_result = manager.start(["--logdir=./logs", "--port=0"])
+    self.assertIsInstance(start_result, manager.StartFailed)
+    self.assertEqual(
+        start_result,
+        manager.StartFailed(
+            exit_code=0,
+            stderr="info: something good happened\n",
+            stdout="also some standard output\n",
+        ),
+    )
+    self.assertEqual(manager.get_all(), [])
+
+  def test_failure_unreadable_stdio(self):
+    if os.name == "nt":
+      # TODO(@wchargin): This could in principle work on Windows.
+      self.skipTest("Requires a POSIX shell for the stub script.")
+    self._stub_tensorboard(
+        name="fail-and-nuke-tmp",
+        program=textwrap.dedent(
+            r"""
+            #!/bin/sh
+            rm -r %s
+            exit 22
+            """.lstrip() % pipes.quote(self.tmproot),
+        ),
+    )
+    start_result = manager.start(["--logdir=./logs", "--port=0"])
+    self.assertIsInstance(start_result, manager.StartFailed)
+    self.assertEqual(
+        start_result,
+        manager.StartFailed(
+            exit_code=22,
+            stderr=None,
+            stdout=None,
+        ),
+    )
+    self.assertEqual(manager.get_all(), [])
+
+  def test_timeout(self):
```

Comment:
what happens if timeout is negative? 🙃

Reply:
Heh. As written, if timeout is negative, it’ll always time out; it won’t
even atempt to read the info files to see if the process launched.

This behavior seems fine enough to me, but if you prefer I’d be happy to
raise `ValueError` if the timeout is negative (and test for that).

Reply:
I was poking to see if I will get a test (without `ValueError`) but not sure how interesting it is.